core/trino-main/src/main/java/io/trino/operator/output/PositionsAppenderPageBuilder.java

```diff
@@ -144,6 +145,31 @@ public boolean isEmpty()
         return declaredPositions == 0;
     }
 
+    public Optional<Page> maybeFlushBeforeRelease()
+    {
+        if (declaredPositions == 0) {
+            return Optional.empty();
+        }
+        boolean forceFlush = false;
+        for (UnnestingPositionsAppender positionsAppender : channelAppenders) {
+            if (positionsAppender.shouldForceFlushBeforeRelease()) {
+                // dictionary encoding will be preserved, so force the current page to be flushed
+                forceFlush = true;
+                break;
+            }
+            else {
+                // transition from dictionary to direct mode if necessary, since we won't be able to reuse the
```

Comment:
Why to do it here? It can be lazily done by next operator instance. This will cause output tracking issues as flattening might increase `outputSizeInBytes`, which is correctly handled by `io.trino.operator.output.PagePartitioner#partitionPage`

Reply:
When done lazily, we are guaranteed to know that all dictionary mode appenders will be forced to flatten. It _will_ correctly report the output size in that situation, but at the cost of losing the dictionary representation (if profitable).

Another thing to note, is that when dictionary appenders flush _without_ flattening, the output size is _incorrect_ both for `partitionPage` and this new method. `PartitionedOutputOperator` stats are always incorrect in the presence of dictionary mode outputs. I think I can fix this at the same time, but it will need to reorganize the output bytes reporting logic.

Reply:
>  When done lazily, we are guaranteed to know that all dictionary mode appenders will be forced to flatten. It will correctly report the output size in that situation, but at the cost of losing the dictionary representation (if profitable).

If the logic is extracted to separate `flattenStuff` method, then we can measure size "correctly" and don't loose next dictionary, e.g: with code like:
```
long positionsAppendersSizeBefore = getPositionsAppendersSizeInBytes();
flattenStuff
long positionsAppendersSizeAfter = getPositionsAppendersSizeInBytes();
operatorContext.recordOutput(positionsAppendersSizeAfter - positionsAppendersSizeBefore, 0 /* no new positions */);
```

Reply:
That works specifically for this prepare release approach, but doesn't fix the general problem when `partitionPage` flushes dictionaries since those don't get flattened and never had their actual size accounted for. I think I can fix both in this PR, will publish a new revision shortly.